### PR TITLE
fix: Block path traversal attacks via the deployment API for MLflow-Triton

### PR DESCRIPTION
**Pull request by mudit-eng**
#### What does the PR do?
This code change sanitizes the model name in the MLflow-Triton deployment plugin to prevent path traversal.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [x] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Test plan:

Added a unit test and verified locally. Also, tested in CI.

- CI Pipeline ID:
44324776, 45253345

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: #xxx

### CODE REVIEW

**Comment by yinggeh** [comment]
Please use template https://github.com/triton-inference-server/server/blob/main/.github/PULL_REQUEST_TEMPLATE/pull_request_template_internal_contrib.md

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 1]
```diff
@@ -513,6 +516,12 @@ def _validate_flavor(self, flavor):
         if flavor not in self.supported_flavors:
             raise Exception("{} model flavor not supported by Triton".format(flavor))
 
+    def _validate_model_name(self, name):
+        if not name:
+            raise Exception("Please provide a model name for the deployment")
+        if '/' in name or '\\' in name or '..' in name:  # Path traversal protection
+            raise Exception("Path traversal is not allowed in model's name: {}".format(name))
```
Comment (yinggeh) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: A more commonly used is `f"Path traversal is not allowed in model's name: {name}"`

File: qa/L0_mlflow/plugin_test.py [hunk 2]
```diff
@@ -116,6 +115,62 @@ def test_onnx_flavor_with_files(self):
             filecmp.cmp(config_path, "./models/onnx_model_with_files/config.pbtxt")
         )
 
+    def test_path_traversal_model_name(self):
```
Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]:
```suggestion
    def test_invalid_path_traversal_model_name(self):
```

File: qa/L0_mlflow/plugin_test.py [hunk 3]
```diff
@@ -116,6 +115,62 @@ def test_onnx_flavor_with_files(self):
...
+
+        model_uri = "models:/onnx_model_with_files/1"
+
+        model_name_normal = "onnx_model_123"
+        self.client_.create_deployment(model_name_normal, model_uri, flavor="onnx")
```
Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]: Why you creating then deleting a deployment?

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 4]
```diff
@@ -513,6 +516,12 @@ def _validate_flavor(self, flavor):
         if flavor not in self.supported_flavors:
             raise Exception("{} model flavor not supported by Triton".format(flavor))
 
+    def _validate_model_name(self, name):
+        if not name:
+            raise Exception("Please provide a model name for the deployment")
+        if '/' in name or '\\' in name or '..' in name:  # Path traversal protection
```
Comment (yinggeh) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: Why isn't `\\` allowed?

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 1 above)
Comment (mudit-eng) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: I will change it. I just followed the convention in this file where .format is used.

File: qa/L0_mlflow/plugin_test.py (hunk 3 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: Earlier, my unit test was failing. I wanted to rule out a potential resource leak, so I added a delete step after a successful create. I don't think it is needed.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 4 above)
Comment (mudit-eng) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: This is for path traversal in Windows. Since we don't support Windows, I will remove it.

Comment (yinggeh) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: What's double backsplash `\\` on windows?

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 1 above)
Comment (yinggeh) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: I see. In this file it's fine. I've seen the other way more commonly used in our code.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 4 above)
Comment (mudit-eng) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: This translates to backslash traversal. Something like this: C:\Windows\System32

Comment (yinggeh) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: Got it. The first backslash is the escape character.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 1 above)
Comment (whoisj) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]: for the record the `\` character is a legal character on POSIX systems.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 5]
```diff
@@ -513,6 +516,14 @@ def _validate_flavor(self, flavor):
         if flavor not in self.supported_flavors:
             raise Exception("{} model flavor not supported by Triton".format(flavor))
 
+    def _validate_model_name(self, name):
```
Comment (yinggeh) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:519]: Just a note no need to fix today, there are multiple similar pattern @whoisj added to address PSIRT issues in server. We could refactor them into one common utility fuction.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 6]
```diff
@@ -99,6 +99,9 @@ def create_deployment(self, name, model_uri, flavor=None, config=None):
         """
         self._validate_flavor(flavor)
 
+        # Validate model name
+        self._validate_model_name(name)
+
         # Verify model does not already exist in Triton
```
Comment (yinggeh) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:105]: This check should also be moved into`_validate_model_name`.

File: qa/L0_mlflow/plugin_test.py [hunk 7]
```diff
@@ -116,6 +116,41 @@ def test_onnx_flavor_with_files(self):
             filecmp.cmp(config_path, "./models/onnx_model_with_files/config.pbtxt")
         )
 
+    def test_invalid_path_traversal_model_name(self):
```
Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]: Not sure if it works but can you try pytest decorator https://github.com/triton-inference-server/server/blob/0844c371fe62c9fb8da9e47e7d76808bdd0ccc58/python/openai/tests/test_lora.py#L43-L56 to make the test cleaner?

Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]:
```suggestion
    def test_invalid_model_name(self):
```

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 6 above)
Comment (mudit-eng) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:105]: Which check should move into `_validate_model_name`?

File: qa/L0_mlflow/plugin_test.py (hunk 7 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: I refactored the code to use the pytest decorator. While the test change itself is straightforward, it also requires environment updates. We would need to modify the test.sh file to make it work with pytest instead of unittest. This increases the scope of the change, so I recommend keeping it as a separate update.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py (hunk 6 above)
Comment (yinggeh) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:105]: I was referring to L105-110. On the second thought, either should be fine. Your call it.
```
        # Verify model does not already exist in Triton
        if self._model_exists(name):
            raise Exception(
                "Unable to create deployment for name %s because it already exists."
```

File: qa/L0_mlflow/plugin_test.py (hunk 7 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: I have made this change

Comment (yinggeh) [reply qa/L0_mlflow/plugin_test.py]: Fair enough. Can you add a "TODO" comment linking to a linear ticket?

Comment (yinggeh) [reply qa/L0_mlflow/plugin_test.py]: Actually, can you refactor into a simple for-loop instead of using pytest?

File: qa/L0_mlflow/plugin_test.py [hunk 8]
```diff
@@ -116,6 +116,41 @@ def test_onnx_flavor_with_files(self):
...
+        model_uri = "models:/onnx_model_with_files/1"
+
+        model_name_empty = ""
```
Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]: Please also test `model_name_empty = "  "`

File: qa/L0_mlflow/plugin_test.py (hunk 7 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: I don't think we need a Linear ticket for pytest. This file is rarely touched and there is no real impact of converting this to pytest. We have other important areas in Triton to work on. If I create a ticket also, TBH, it will just remain in the backlog.
For the for-loop suggestion, what simplification are you looking for - readability or maintainability or something else? We just have 3 cases to tests and code is straightforward. And as we know, this file is rarely touched. I don't think it is worth spending time going back and forth with pytest decorators or for-loop, unless there is a real benefit.

File: qa/L0_mlflow/plugin_test.py (hunk 8 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: What would be the reason for testing 'spaces'? Spaces are allowed in the model name. If we go in this direction, then we can test lot of cases like ".", "#", "      ", and so on. Scope of this PR is to do security hardening by preventing path traversal.

Comment (yinggeh) [reply qa/L0_mlflow/plugin_test.py]: Because we've seen server error with empty model names. The point of tests is not testing that what we know will work. The point of tests is to discover as many corner cases before thery are reported by our customers and any orgs, like what you have spent time fixing now. It also prevents anyone in the future who accidentally touchs this code and causes regression.

> then we can test lot of cases like ".", "#", " ", and so on

And yes, I don't mind having more test cases if they are necessary. See how @pskiran1 did https://github.com/triton-inference-server/server/pull/8672/changes#diff-46efca2b90e4866165c6eda38961abdc21ab0bbcbe81f1dfa916489139fc08fe

File: qa/L0_mlflow/plugin_test.py (hunk 7 above)
Comment (yinggeh) [reply qa/L0_mlflow/plugin_test.py]: I don’t see functional risk in keeping this as-is. My only suggestion is around long-term maintainability — parameterized tests tend to scale better when validating path validation logic like this. This could reduce duplication and improve scalability if more cases are added later. If we expect this area to evolve, it might be worth adopting that pattern now while the test surface is small.

File: qa/L0_mlflow/plugin_test.py (hunk 8 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: I totally understand what you are saying @yinggeh .

I can add non-path traversal cases, but this PR is primarily focused on blocking path traversal as reported in PSIRT. We can expand the scope of the ticket to include model name validation if needed. I just want to be mindful of potential scope creep. 

I have seen Sai’s PR. @pskiran1 is still testing path traversal scenarios. Although he has added test cases covering spaces, non-printable characters, etc., he is still validating for path traversal patterns (.., /) within those cases.
I don't think he is explicitly testing `model_name_empty = "  "`.

What are other invalid model names I can test with? Are there other characters that are not allowed? Let's have a list so that both PRs (Sai's and mine) can include those.

@whoisj

File: qa/L0_mlflow/plugin_test.py (hunk 7 above)
Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: Fair point. Adding for-loop should be trivial. I can do that. But I don't think we should go to pytest direction for the same reasons I mentioned earlier.

File: qa/L0_mlflow/plugin_test.py (hunk 8 above)
Comment (whoisj) [reply qa/L0_mlflow/plugin_test.py]: I'm 100% fine with keeping the scope of this PR being kept minimal, so long as we create a tracking ticket internally for the work being put off for a future PR.

Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: I have made the change to check that model name is not empty and does not only contains whitespace, tabs, or newlines.
For any other validation, we can have a tracking ticket.

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py [hunk 9]
```diff
@@ -513,6 +516,16 @@ def _validate_flavor(self, flavor):
         if flavor not in self.supported_flavors:
             raise Exception("{} model flavor not supported by Triton".format(flavor))
 
+    def _validate_model_name(self, name):
+        # Check if the model name is empty or only contains whitespace, tabs, or newlines
+        if name.strip() == "":
+            raise Exception("Please provide a valid model name for the deployment")
```
Comment (yinggeh) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py]:
```suggestion
            raise Exception("Empty model name is not allowed. Please provide a valid model name for the deployment")
```

File: qa/L0_mlflow/plugin_test.py [hunk 10]
```diff
@@ -116,6 +116,66 @@ def test_onnx_flavor_with_files(self):
...
+                str(e.exception),
+            )
+            self.assertNotIn(
+                f"Path traversal is not allowed in model's name: {model_name}",
+                str(e.exception),
+            )
+            self.assertIn(
+                f"Registered Model with name={model_name} not found",
+                str(e.exception),
+            )
```
Comment (yinggeh) [inline qa/L0_mlflow/plugin_test.py]:
```suggestion
            self.client_.create_deployment(model_name, model_uri, flavor="onnx")
```

**Review by yinggeh — CHANGES_REQUESTED** [review]

File: qa/L0_mlflow/plugin_test.py (hunk 10 above)
Comment (yinggeh) [reply qa/L0_mlflow/plugin_test.py]: The test fails if exception is raised.

Comment (mudit-eng) [reply qa/L0_mlflow/plugin_test.py]: Rewrote the code to change asserts.

**Review by yinggeh — APPROVED** [review]
Nice work
